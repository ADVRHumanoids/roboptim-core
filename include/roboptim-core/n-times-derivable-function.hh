// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class NTimesDerivable declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_N_TIMES_DERIVABLE_HH
# define ROBOPTIM_TRAJECTORY_N_TIMES_DERIVABLE_HH
# include <boost/static_assert.hpp>

# include <roboptim-core/twice-derivable-function.hh>

namespace roboptim
{
  /**
     \addtogroup roboptim_function
     @{
  */

  template <unsigned dorder>
  class NTimesDerivableFunction;

  template <>
  class NTimesDerivableFunction<2> : public TwiceDerivableFunction
  {
  public:
    static const size_type derivabilityOrder = 2;

    NTimesDerivableFunction (size_type m = 1) throw ()
      : TwiceDerivableFunction (1, m)
    {}

    virtual ~NTimesDerivableFunction () throw ()
    {}

    virtual vector_t operator () (const vector_t& x) const throw ()
    {
      assert (x.size () == 1);
      return operator () (x[0]);
    }

    virtual vector_t operator () (double) const throw () = 0;

    virtual gradient_t gradient (const vector_t& x, int i = 0) const throw ()
    {
      assert (x.size () == 1);
      vector_t d = derivative (x[0], 1);

      vector_t res (1);
      res[0] = d[i];
      return res;
    }

    virtual hessian_t hessian (const vector_t& x, int i) const throw ()
    {
      hessian_t hessian (1, 1);

      vector_t dd = derivative (x[0], 2);
      hessian (0, 0) = dd[i];

      return hessian;
    }

    virtual vector_t derivative (double x, size_type order = 1) const
      throw () = 0;

    virtual std::ostream& print (std::ostream& o) const throw ()
    {
      o << "Function derivable " << derivabilityOrder << " times.";
      return o;
    }

  };

  /// \brief Define a \f$\mathbb{R} \rightarrow \mathbb{R}^m\f$ function,
  /// derivable n times (n >= 2).
  template <unsigned DerivabilityOrder>
  class NTimesDerivableFunction
    : public NTimesDerivableFunction<DerivabilityOrder - 1>
  {
  public:
    BOOST_STATIC_ASSERT(DerivabilityOrder > 2);

    typedef Function::size_type size_type;
    static const size_type derivabilityOrder = DerivabilityOrder;

    NTimesDerivableFunction (size_type m = 1) throw ()
    : NTimesDerivableFunction<DerivabilityOrder - 1> (m)
    {
    }

    virtual ~NTimesDerivableFunction () throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  };

  /**
     @}
  */
}
# include <roboptim-core/n-times-derivable-function.hxx>
#endif //! ROBOPTIM_TRAJECTORY_N_TIMES_DERIVABLE_HH
