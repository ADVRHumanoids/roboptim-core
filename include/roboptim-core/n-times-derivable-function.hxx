// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class NTimesDerivable implementation.
 */

#ifndef ROBOPTIM_CORE_N_TIMES_DERIVABLE_HXX
# define ROBOPTIM_CORE_N_TIMES_DERIVABLE_HXX

namespace roboptim
{
  template <unsigned dorder>
  NTimesDerivableFunction<dorder>::NTimesDerivableFunction (size_type m)
    throw ()
    : TwiceDerivableFunction (1, m)
  {
  }

  template <unsigned dorder>
  NTimesDerivableFunction<dorder>::~NTimesDerivableFunction () throw ()
  {
  }

  template <unsigned dorder>
  typename NTimesDerivableFunction<dorder>::vector_t
  NTimesDerivableFunction<dorder>::operator () (const vector_t& x) const
    throw ()
  {
    assert (x.size () == 1);
    return operator () (x[0]);
  }

  template <unsigned dorder>
  typename NTimesDerivableFunction<dorder>::gradient_t
  NTimesDerivableFunction<dorder>::gradient (const vector_t& x, int i) const
    throw ()
  {
    assert (x.size () == 1);
    vector_t d = derivative (x[0], 1);

    vector_t res (1);
    res[0] = d[i];
    return res;
  }

  template <unsigned dorder>
  typename NTimesDerivableFunction<dorder>::hessian_t
  NTimesDerivableFunction<dorder>::hessian (const vector_t& x, int i)
    const throw ()
  {
    hessian_t hessian (1, 1);

    vector_t dd = derivative (x[0], 2);
    hessian (0, 0) = dd[i];

    return hessian;
  }

  template <unsigned dorder>
  std::ostream&
  NTimesDerivableFunction<dorder>::print (std::ostream& o) const
    throw ()
  {
    o << "Function derivable " << derivabilityOrder << " times.";
    return o;
  }
}

#endif //! ROBOPTIM_CORE_N_TIMES_DERIVABLE_HXX
