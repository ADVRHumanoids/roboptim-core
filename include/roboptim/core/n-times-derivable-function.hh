// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Class NTimesDerivable declaration.
 */

#ifndef ROBOPTIM_TRAJECTORY_N_TIMES_DERIVABLE_HH
# define ROBOPTIM_TRAJECTORY_N_TIMES_DERIVABLE_HH
# include <boost/static_assert.hpp>

# include <roboptim/core/twice-derivable-function.hh>

namespace roboptim
{
  /**
     \addtogroup roboptim_function
     @{
  */

  template <unsigned dorder>
  class NTimesDerivableFunction;

  template <>
  class NTimesDerivableFunction<2> : public TwiceDerivableFunction
  {
  public:
    static const size_type derivabilityOrder = 2;


    virtual ~NTimesDerivableFunction () throw ()
    {}


    /// Return the size of the derivative vector.
    size_type derivativeSize () const throw ()
    {
      return outputSize ();
    }

    /// Check if a derivative is valid (check sizes).
    bool isValidDerivative (const gradient_t& derivative) const throw ()
    {
      return derivative.size () == this->derivativeSize ();
    }


    /// Evaluate the function at a given point.
    result_t operator () (double argument) const
      throw ()
    {
      result_t result (outputSize ());
      (*this) (result, argument);
      return result;
    }


    /// Evaluate the function at a given point.
    void operator () (result_t& result, double argument) const throw ()
    {
      assert (isValidResult (result));
      this->impl_compute (result, argument);
      assert (isValidResult (result));
    }


    /// Compute the derivative of the function at a certain order
    /// and for a given point.
    gradient_t derivative (double argument, size_type order = 1) const
      throw ()
    {
      gradient_t derivative (derivativeSize ());
      this->derivative (derivative, argument, order);
      return derivative;
    }


    /// Compute the derivative of the function at a certain order
    /// and for a given point.
    void derivative (gradient_t derivative,
		     double argument,
		     size_type order = 1) const
      throw ()
    {
      assert (order <= derivabilityOrder
	      && isValidDerivative (derivative));
      this->impl_derivative (derivative, argument, order);
      assert (isValidDerivative (derivative));
    }


    virtual std::ostream& print (std::ostream& o) const throw ()
    {
      o << "Function derivable " << derivabilityOrder << " times.";
      return o;
    }

  protected:
    /// Constructor.
    NTimesDerivableFunction (size_type outputSize = 1) throw ()
      : TwiceDerivableFunction (1, outputSize)
    {}

    /// Implement the generic evaluation method, as required by the Function
    /// class, using the specific operator () method declared in this class.
    void impl_compute (result_t& result, const argument_t& argument)
      const throw ()
    {
      (*this) (result, argument[0]);
    }

    /// Evaluate the function, has to be implemented by the concrete class.
    virtual void impl_compute (result_t&, double) const throw () = 0;

    /// Implement the generic gradient computation method, as required by
    /// the DerivableFunction class, using the derivative method declared
    /// in this class.
    void impl_gradient (gradient_t& gradient,
			const argument_t& argument,
			int functionId = 0) const throw ()
    {
      assert (functionId == 0);
      this->derivative (gradient, argument[0], 1);
    }


    /// Derivative computation, has to be implemented in concrete class.
    virtual void impl_derivative (gradient_t& derivative,
				  double argument,
				  size_type order = 1) const throw () = 0;

    void impl_hessian (hessian_t& hessian,
		       const argument_t& argument,
		       int functionId = 0) const throw ()
    {
      assert (functionId == 0);

      gradient_t gradient (gradientSize ());
      this->derivative (gradient, argument[0], 2);

      hessian (0, 0) = gradient[functionId];
    }
  };

  /// \brief Define a \f$\mathbb{R} \rightarrow \mathbb{R}^m\f$ function,
  /// derivable n times (n >= 2).
  template <unsigned DerivabilityOrder>
  class NTimesDerivableFunction
    : public NTimesDerivableFunction<DerivabilityOrder - 1>
  {
  public:
    BOOST_STATIC_ASSERT(DerivabilityOrder > 2);

    typedef Function::size_type size_type;
    static const size_type derivabilityOrder = DerivabilityOrder;

    virtual ~NTimesDerivableFunction () throw ();

    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    /// Constructor.
    NTimesDerivableFunction (size_type m = 1) throw ()
      : NTimesDerivableFunction<DerivabilityOrder - 1> (m)
    {}
  };

  /**
     @}
  */
}
# include <roboptim/core/n-times-derivable-function.hxx>
#endif //! ROBOPTIM_TRAJECTORY_N_TIMES_DERIVABLE_HH
