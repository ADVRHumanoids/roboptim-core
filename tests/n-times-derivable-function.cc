// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#undef NDEBUG

#include "shared-tests/fixture.hh"

#include <roboptim/core/io.hh>
#include <roboptim/core/n-times-derivable-function.hh>

using namespace roboptim;

// Define a 10 times derivable null function.
struct F : public NTimesDerivableFunction<10>
{
  using NTimesDerivableFunction<10>::impl_compute;

  F () : NTimesDerivableFunction<10> (4, "0")
  {}

  virtual void impl_compute (result_t& result, double) const
  {
    result.setZero ();
  }

  virtual void impl_derivative (gradient_t& derivative,
				double,
				size_type order = 1) const
  {
    assert (order <= derivabilityOrder);
    derivative.setZero ();
  }
};

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (n_times_derivable_function)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("n-times-derivable-function");

  boost::shared_ptr<F> fct = boost::make_shared<F> ();
  F::argument_t x (1);
  x.setZero ();

  (*output) << (*fct) << std::endl
	    << "Evaluate:" << std::endl
	    << (*fct) (x[0]) << std::endl
	    << "Gradient:" << std::endl
	    << fct->gradient (x, 0) << std::endl
	    << "Jacobian:" << std::endl
	    << fct->jacobian (x) << std::endl;

  BOOST_CHECK (fct->derivabilityOrderMax () == 10);

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
