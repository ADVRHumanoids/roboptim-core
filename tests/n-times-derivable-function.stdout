10-time differentiable function.
Evaluate:
[4](0,0,0,0)
Gradient:
[1](0)
Jacobian:
[4](0,0,0,0)

